dnsmasq: add filter-rr and cache-rr options

systemcrash opened the pull request:
Enable the `--filter-rr=...`, `--cache-rr=...` and `--dns-rr=...` options.

Also added handling of a separate config file included in the `/tmp/dnsmasq[.instance].d` which is a future-proofing to handle new dnsmasq options which the init file or GUI does not yet.

systemcrash committed:
dnsmasq: add handling of `dns-rr` to init script (add arbitrary resource records)

Add support for handling of DNS RR (Resource Records) requests, which
are needed for the HTTPS Type 65 records, introduced to support the
DNS-based Service Discovery (DNS-SD) mechanism for HTTPS services and
defined in the RFC 9460 (9.1. Query Names for HTTPS RRs).

Ref: https://forum.openwrt.org/t/resolving-query-type-65-to-local-address-for-ios-clients-in-dnsmasq/179504/11

uci config usage:

config dnsrr
    option rrname 'foo.example.com'
    option rrnumber '65'
    option hexdata '00'

hexdata is optional.

Available since dnsmasq 2.62 (for around 12 years at this point).

Note: dnsmasq dns-rr are not affected by filter-rr

Tested on 22.03.5

Signed-off-by: Paul Donald <newtwen+github@gmail.com>
Tested-by: Vladimir Kochkovski <ask@getvladimir.com>
Link: https://github.com/openwrt/openwrt/pull/14975
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

systemcrash committed:
dnsmasq: add handling of `cache-rr` to init script

dnsmasq v2.90 introduced `--cache-rr=<rrtype>[,<rrtype>...]`.

uci config usage:

config dnsmasq
    ...
    option cache_rr 'AAAA,CNAME,NXDOMAIN,SRV,...'

The dnsmasq instance internally builds a linked list of RR to cache
from the individually supplied parameters, so it's allowed to provide
multiples:

... --cache-rr=AAAA --cache-rr=A ...

See https://forum.openwrt.org/t/resolving-query-type-65-to-local-address-for-ios-clients-in-dnsmasq/179504

Tested on: 23.05.2

Signed-off-by: Paul Donald <newtwen+github@gmail.com>
Tested-by: Vladimir Kochkovski <ask@getvladimir.com>
Link: https://github.com/openwrt/openwrt/pull/14975
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

systemcrash committed:
dnsmasq: quoted path variables

Prevents problems when variables contain spaces.

Tested on: 23.05.3

Signed-off-by: Paul Donald <newtwen+github@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14975
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

systemcrash commented:
@pprindeville you always seem to be able to find the holes. Do you see anything wrong with this commit included here?

https://github.com/openwrt/openwrt/pull/14975/commits/ab46c2201a5662cb94f6a6e38f7143045c96811e

systemcrash committed:
dnsmasq: add handling of `filter-rr` to init script

dnsmasq v2.90 introduced `--filter-rr=<rrtype>[,<rrtype>...]`.

uci config usage:

config dnsmasq
    ...
    option filter_rr 'AAAA,CNAME,NXDOMAIN,SRV,...'

The dnsmasq instance internally builds a linked list of RR to filter
from the individually supplied parameters, so it's harmless to provide
synonyms:

... --filter-A --filter-rr=A ...

See https://forum.openwrt.org/t/resolving-query-type-65-to-local-address-for-ios-clients-in-dnsmasq/179504/23

Tested on: 23.05.2

Signed-off-by: Paul Donald <newtwen+github@gmail.com>
Tested-by: Vladimir Kochkovski <ask@getvladimir.com>
Link: https://github.com/openwrt/openwrt/pull/14975
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

systemcrash commented:
Can we get a review/approval @nbd168 @ynezz or other devs - this is largely the same as #14269

jow- commented:
Why was https://github.com/openwrt/openwrt/pull/14269 closed? What's the difference compared to this PR?

systemcrash commented:
@jow- 
> Why was #14269 closed? What's the difference compared to this PR?

More commits, especially the 'custom dnsmasq config' (per instance) which might prevent future commits.

pprindeville commented:
> @pprindeville you always seem to be able to find the holes. Do you see anything wrong with this commit included here?
> 
> [ab46c22](https://github.com/openwrt/openwrt/commit/ab46c2201a5662cb94f6a6e38f7143045c96811e)

Sorry for just looking at this now...

pprindeville commented:
> @jow-
> 
> > Why was #14269 closed? What's the difference compared to this PR?
> 
> More commits, especially the 'custom dnsmasq config' (per instance) which might prevent future commits.

I would have done it on the same PR just for continuity...

pprindeville commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -787,6 +788,28 @@ dhcp_hostrecord_add() {
+
+	# dnsmasq accepts colon XX:XX:.., space XX XX .., or contiguous XXXX.. hex forms or mixtures thereof
+	if [ -n "${hexdata//[0-9a-fA-F\:\ ]/}" ]; then
+		# is invalid hex literal
```
I'd spew something like the `trust-anchor` handling, i.e.:

```
echo "dnsmasq: \"$hexdata\" is malformed hexadecimal string" >&2
return 1
```

instead.

systemcrash replied:
Good idea. Taken.

pprindeville commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -1115,13 +1141,18 @@ dnsmasq_start()
 
-	config_get dnsmasqconfdir "$cfg" confdir "/tmp/dnsmasq.d"
+	# Create a dnsmasq.d dir for each instance
+	config_get dnsmasqconfdir "$cfg" confdir "/tmp/dnsmasq${cfg:+.$cfg}.d"
```
Do you need to test it as being an absolute path, i.e. `"${dnsmasqconfdir:0:1}" = '/'` ?

systemcrash replied:
Just doing what original author did. Or did you have something like chroot or jail in mind?

pprindeville replied:
No, just trying to avoid ambiguity if someone thinks it’s a relative directory and it’s not.

systemcrash replied:
Could you give me an example that you think might be problematic?

pprindeville replied:
What if they have:

```
option confdir “..”
```

systemcrash replied:
OK. Got it. I suppose a user could do that. There does not seem to be harm in doing so, and that's possible today. The resulting path depends on how dnsmasq and the init script interpret it. Those will likely differ, and cause weird things to happen. 


We can test for absolute as your test. What are we thinking, gate all of this conf stuff behind e.g. this?
```sh
	config_get dnsmasqconfdir "$cfg" confdir "/tmp/dnsmasq${cfg:+.$cfg}.d"
	[ "${dnsmasqconfdir:0:1}" = '/' ] && {
	# Stuff
	}
```

systemcrash replied:
Looking at later lines, like `		dnsmasqconfdir="${dnsmasqconfdir%%,*}"` it seems that it maybe once upon a time accepted `dir1,dir2,dir3`.

pprindeville replied:
So, an iterator is needed? Or a `list` instead?

systemcrash replied:
I don't know. I think whoever wrote it gave up on the idea and settled on the first entry. Happy to let sleeping dogs lie...

pprindeville commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -1115,13 +1141,18 @@ dnsmasq_start()
+
+	# EXTRACONFFILE allows new dnsmasq parameters before they are natively handled in this init file
+	config_get extraconftext "$cfg" extraconftext
+	echo -e "$extraconftext" > "$dnsmasqconfdir"/"$EXTRACONFFILE"
```
```
[ -n "$extraconftext" ] && ...
```

systemcrash replied:
Good point ☝️ Taken

pprindeville reviewed (changes requested):
Minor comments...  I'm liking the quoting of variables expanding into path names...

systemcrash committed:
dnsmasq: add 'extraconftext' parameter

Users can now freely add new dnsmasq parameters (i.e. a whole config)
via extraconf. This means users can add their own parameters without
changes to init or GUI.

Co-opted the default of confdir also to include the instance name.
This way each instance gets its own .d directory (and separate instances
do not all inherit the same 'extraconftext').

Usage:
config dnsmasq 'config'
	...
	option extraconftext 'cache-size=2048\nlog-async=20'

config dnsmasq 'blah'
	...
	option extraconftext 'cache-size=128\nlog-async=5'

or even (which would produce staggered output but still valid)

config dnsmasq 'blah'
	...
	option extraconftext 'cache-size=128
							log-async=5'

See https://forum.openwrt.org/t/add-dnsmasq-custom-options-field-in-luci-gui/193184

Tested on: 23.05.3, 22.03.6

Signed-off-by: Paul Donald <newtwen+github@gmail.com>
Tested-by: Vladimir Kochkovski <ask@getvladimir.com>
Link: https://github.com/openwrt/openwrt/pull/14975
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

systemcrash committed:
dnsmasq: gate configdir usage behind absolute path check

don't use configuration directories which are relative

Signed-off-by: Paul Donald <newtwen+github@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14975
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

egc112 reviewed (approved):
Tested on DL-WRX36 main K6.6
Works as advertised.
I used the example for extraconfig (option extraconftext 'cache-size=128`)
this resulted in
```
Mon Apr 22 15:39:04 2024 daemon.crit dnsmasq[1]: illegal repeated keyword at line 1 of /tmp/dnsmasq.d/extraconfig.conf
Mon Apr 22 15:39:04 2024 daemon.crit dnsmasq[1]: FAILED to start up
Mon Apr 22 15:39:04 2024 daemon.info procd: Instance dnsmasq::cfg01411c s in a crash loop 6 crashes, 0 seconds since last crash
```
The config already had this option by default so perhaps choose another example.
But after I removed the default `option cachesize '1000'` everything was working
Thanks for all your work

systemcrash commented:
Interesting. Maybe a patch to dnsmasq is required to perhaps silently ignore cases like these or warn instead. Would need to read further.

schuettecarsten commented:
Any plans to patch Luci for these new options?

GetVladimir commented:
@schuettecarsten Ideally yes, here is an example of how the Dnsmasq Custom Options field might look in the GUI/LuCi: https://forum.openwrt.org/t/add-dnsmasq-custom-options-field-in-luci-gui/193184

Quite a few Dnsmasq options have been waiting for a long time to be reviewed and implemented in uci and LuCi, so hopefully the Custom Options filed will help to address this.

If it works good, at least we won't need to make a manual patch for each new Dnsmasq feature and wait to be implemented, which should speed up the whole process and free up time and resources for other updates

systemcrash commented:
> Any plans to patch Luci for these new options? This patch for EDNS options is still pending: [openwrt/luci#6972](https://github.com/openwrt/luci/pull/6972) It waits for #14349, which is not reviews/completed yet.

The original submitter has showed no interest to fix the code in the PR you linked. It might go quicker if you make a new PR which corrects those errors, and flag the original PR to be closed.

systemcrash commented:
ping

pprindeville reviewed (approved):
LGTM

edrikk reviewed (approved):
LGTM.
Very much needed.  I like to avoid touching the out of box files such as dnsmasq.conf as much as possible.

GetVladimir commented:
@edrikk thank you so much for the review and for the comment. You're right and I agree regarding keeping the out of box files to defaults whenever possible.

Hopefully the solution will get added for everyone to benefit

systemcrash commented:
Any final reviews?

stangri commented:
[These lines](https://github.com/openwrt/openwrt/commit/bd81d97e19e6cc6e33dc5ff852ece95bbc6be01e#diff-2d0292e13a91f53bc368201bcb996f829138e0ad102f0249df523998b5383549L1124-R1126) cause an issue.

Since dnsmasq is no longer including the default "/tmp/dnsmasq.d" config in each instance, this is breaking some of the packages relying on dnsmasq-full, like pbr, some configurations of adblock-fast, possibly others (adblock).

In the scenario of using dnsmasq config files for ad-blocking for example, what is the recommended way to set it for multiple instances after this change? Copy a relatively big block-list into each instance's config dir?

I think each instance should include the default "/tmp/dnsmasq.d" and the per-instance "/tmp/dnsmasq${cfg}.d".

@systemcrash @hauke please advise if you want me to create a PR to include both directories and fix this.

systemcrash commented:
Hmm. For a start, I don't find it good practice to expect directories to be there unless the package itself 'owns' that directory. 

Does it work to ask uci for that instances `confdir`? That seems the most sensible way.

The problem with having the main directory is filtering out edits which should only go into the instance config.

systemcrash commented:
> Copy a relatively big block-list into each instance's config dir?

Sym link? Or.. sym link to a dir? Not sure whether the latter would get processed tho.

systemcrash commented:
FWIW github hangs trying to load your comment on the actual commit... good you posted here too.

stangri commented:
> Sym link? Or.. sym link to a dir? Not sure whether the latter would get processed tho.

I haven't browsed the most recent dnsmasq init-file, but links may not work if the instance is jailed to a specific configdir. 

> Hmm. For a start, I don't find it good practice to expect directories to be there unless the package itself 'owns' that directory.

It's not about directory being there, it's about the directory which was previously used for all dnsmasq instances confdirs no longer being used.

Adding files to `/tmp/dnsmasq.d/` is a standard practice for other packages integrating into dnsmasq. Having to duplicate these files into each instance's own confdir, especially when it's a 10Mb-large ab-blocking file doesn't seem practical.


> The problem with having the main directory is filtering out edits which should only go into the instance config.

Don't use the main directory's file in your own configs then.

systemcrash commented:
> It's not about directory being there, it's about the directory which was previously used for all dnsmasq instances confdirs no longer being used.

The main instance is still there - it now assumes the name of its config instance. 

Perhaps you have a suggestion to reconcile?

> Don't use the main directory's file in your own configs then.

In some cases you must: otherwise how would the main config get exclusive extraconfig edits without other instances inheriting? If each directory is chained to its instance (id), then this problem is resolved.

stangri commented:
> The main instance is still there - it now assumes the name of its config instance.

I do not suggest this PR is rolled back, I suggest dnsmasq is set to include both per-instance config dirs and the default config dir (if I'm not mistaken, you can instruct dnsmasq to include multiple directories for config).

I don't see any drawbacks of the suggested solution. If you want per-instance config files, use the per-instance directories, if you want some files to affect all instances, you place them into `/tmp/dnsmasq.d/`.

gSpotx2f commented:
>Having to duplicate these files into each instance's own confdir, especially when it's a 10Mb-large ab-blocking file doesn't seem practical.

>I don't see any drawbacks of the suggested solution. If you want per-instance config files, use the per-instance directories, if you want some files to affect all instances, you place them into /tmp/dnsmasq.d/.

What about RAM usage? When each dnsmasq instance will load a large file from /tmp/dnsmasq.d/ into itself.

stangri commented:
> > 
> 
> What about RAM usage? When each dnsmasq instance will load a large file from /tmp/dnsmasq.d/ into itself.

What about RAM usage when on top of each dnsmasq instance loading a large file, it also has to be in RAM in "/tmp/dnsmasq${cfg}.d"?

stangri commented:
On top of `pbr`, here's a list of packages from openwrt/packages repo which depend on the default confdir being `/tmp/dnsmasq.d/`: https://github.com/search?q=repo%3Aopenwrt%2Fpackages%20%2Ftmp%2Fdnsmasq&type=code
